### PR TITLE
Fnymmm/save secret

### PR DESCRIPTION
**Pull request by fnymmm**
Added backend for "save secret" toggle switch in api-resource tab

- Added EncryptedSecret and EncryptionKey properties
- Connects to Azure KeyVault to encrypt secrets

### CODE REVIEW

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        _keyVaultUri = options.KeyVaultUri;
+        _keyVaultKeyName = options.EncryptionKeyName;
+        _cryptographyClient = GetCryptographyClient().Result;
+        EncryptionAlgorithm = EncryptionAlgorithm.RsaOaep;
+    }
+
+    public string GetEncryptionKeyNameBase64()
+    {
+        var nameAsArray = Encoding.UTF8.GetBytes(_keyVaultKeyName);
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:33]: we can also just create private variable that is already encoded as base64

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    {
+
+
+        _keyVaultUri = options.KeyVaultUri;
+        _keyVaultKeyName = options.EncryptionKeyName;
+        _cryptographyClient = GetCryptographyClient().Result;
+        EncryptionAlgorithm = EncryptionAlgorithm.RsaOaep;
+    }
+
+    public string GetEncryptionKeyNameBase64()
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:31]: Base 64 does not really encrypt data, its more of an encoding

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    private readonly CryptographyClient _cryptographyClient;
+    public EncryptionAlgorithm EncryptionAlgorithm { get; }
+
+    public KeyVaultController(AzureKeyVaultOptions options)
+    {
+
+
+        _keyVaultUri = options.KeyVaultUri;
+        _keyVaultKeyName = options.EncryptionKeyName;
+        _cryptographyClient = GetCryptographyClient().Result;
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:27]: This has to be implemented differently. `.Result` on a Task is a big nono

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        return Encoding.Default.GetString(result.Plaintext);
+    }
+
+    private async Task<CryptographyClient> GetCryptographyClient()
+    {
+        var client = new KeyClient(new Uri(_keyVaultUri), new DefaultAzureCredential());
+        KeyVaultKey key = await client.GetKeyAsync(_keyVaultKeyName).ConfigureAwait(false);
+        return new CryptographyClient(key.Id, new DefaultAzureCredential());
+    }
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:62]: I think we should move this to a factory. Then you can make this a lot cleaner

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+using Azure.Identity;
+using Azure.Security.KeyVault.Keys;
+using Azure.Security.KeyVault.Keys.Cryptography;
+using MassTransit.Configuration;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+
+namespace IdOps.Controller;
+
+public class KeyVaultController : IKeyVaultController
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:13]: I dont think this is a controller

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        var inputAsArray = Encoding.UTF8.GetBytes(input);
+
+        EncryptResult result = await _cryptographyClient
+            .EncryptAsync(EncryptionAlgorithm, inputAsArray).ConfigureAwait(false);
+
+        return Convert.ToBase64String(result.Ciphertext);
+    }
+
+    public async Task<string> Decrypt(string input)
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:47]:
```suggestion
    public async Task<string> DecryptAsync(string input, CancellcationToken cancellationToken)
```

File: src/Server/src/Core/Controllers/KeyVaultController.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        EncryptionAlgorithm = EncryptionAlgorithm.RsaOaep;
+    }
+
+    public string GetEncryptionKeyNameBase64()
+    {
+        var nameAsArray = Encoding.UTF8.GetBytes(_keyVaultKeyName);
+        return Convert.ToBase64String(nameAsArray);
+    }
+
+    public async Task<string> Encrypt(string input)
```
Comment (PascalSenn) [inline src/Server/src/Core/Controllers/KeyVaultController.cs:37]:
```suggestion
    public async Task<string> EncryptAsync(string input, CancellcationToken cancellationToken)
```

**Review by PascalSenn — CHANGES_REQUESTED** [review]

File: src/Server/src/Core/Controllers/IKeyVaultController.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+namespace IdOps.Controller;
+
+public interface IKeyVaultController
+{
+
+    public string GetEncryptionKeyNameBase64();
+
+    public Task<string> Encrypt(string input);
+
+    public Task<string> Decrypt(string input);
```
[the remark this answers is not in the document]
Comment (fnymmm) [reply src/Server/src/Core/Controllers/IKeyVaultController.cs:12]: File was renamed -> added changes to renamed file

File: src/Server/src/Core/Controllers/KeyVaultController.cs (hunk 4 above)
Comment (fnymmm) [reply src/Server/src/Core/Controllers/KeyVaultController.cs:60]: I now inject the CryptographyClient

File: src/Server/src/Encryption.KeyVault/KeyVaultController.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+using Azure.Identity;
+using Azure.Security.KeyVault.Keys;
+using Azure.Security.KeyVault.Keys.Cryptography;
+using EncryptionAlgorithm = Azure.Security.KeyVault.Keys.Cryptography.EncryptionAlgorithm;
+
+
+public class KeyVaultController : IEncryptionService
+{
+    private readonly CryptographyClient _cryptographyClient;
+    public EncryptionAlgorithm EncryptionAlgorithm { get; }
```
Comment (PascalSenn) [inline src/Server/src/Encryption.KeyVault/KeyVaultController.cs:12]: private

File: src/Server/src/Encryption.KeyVault2/Program.cs [hunk 10]
```diff
@@ -0,0 +1,6 @@
+var builder = WebApplication.CreateBuilder(args);
```
Comment (PascalSenn) [inline src/Server/src/Encryption.KeyVault2/Program.cs:1]: remove

File: src/Server/test/Encryption.KeyVault.Tests/KeyVaultControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    /*
+     *  Conversion Chart
+     *
+     *  Foo;    Rm9v;       {70, 111, 111}
+     *  Bar;    QmFy;       {66, 97, 114}
+     *  FooBar; Rm9vQmFy;   {70, 111, 111, 66, 97, 114}
+     *
+     */
+
+
```
Comment (PascalSenn) [inline src/Server/test/Encryption.KeyVault.Tests/KeyVaultControllerTests.cs:19]:
```suggestion
```

File: src/Server/test/Encryption.KeyVault.Tests/KeyVaultControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    [InlineData("Rm9v")]
+    [InlineData("QmFy")]
+    [InlineData("Rm9vQmFy")]
+    public async void DecryptAsync_Should_ReturnString(string input)
+    {
+        //Arrange
+        var inputAsArray = Convert.FromBase64String(input);
+        var decryptResult = CryptographyModelFactory.DecryptResult(plaintext: inputAsArray);
+
+
```
Comment (PascalSenn) [inline src/Server/test/Encryption.KeyVault.Tests/KeyVaultControllerTests.cs:58]:
```suggestion
```

**Review by PascalSenn — APPROVED** [review]

File: src/Server/src/Core/Controllers/KeyVaultController.cs (hunk 4 above)
Comment (fnymmm) [reply src/Server/src/Core/Controllers/KeyVaultController.cs]: implemented provider

**Pull request merged**